### PR TITLE
Exchange && ringbuffer support blf.

### PR DESCRIPTION
**Pull request by xinyi-xs**

### CODE REVIEW

File: include/nng/supplemental/nanolib/conf.h [hunk 1]
```diff
@@ -373,6 +373,17 @@ struct conf_parquet {
 };
 typedef struct conf_parquet conf_parquet;
 
+struct conf_blf {
+	bool                    enable;
+	char	           		*dir;
+	char	           		*file_name_prefix;
```
Comment (wanghaEMQ) [inline include/nng/supplemental/nanolib/conf.h:379]: Seems not formatted

File: src/supplemental/nanolib/blf/blf.cc [hunk 2]
```diff
@@ -249,8 +249,11 @@ blf_write_core(
 			/* write a CanMessage */
 			blf_write_can_message(file, frame);
 		}
+
+        cJSON_Delete(jso);
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/blf/blf.cc:253]: not formatted

File: src/supplemental/nanolib/blf/blf.cc [hunk 3]
```diff
@@ -331,6 +334,10 @@ blf_write_loop(void *config)
 int
 blf_write_batch_async(blf_object *elem)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("BLF is not ready or not launch!");
+        return -1;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/blf/blf.cc:340]: not formatted

File: src/supplemental/nanolib/blf/blf.cc [hunk 4]
```diff
@@ -382,6 +389,10 @@ compare_callback_span(void *name, uint64_t low, uint64_t high)
 const char *
 blf_find(uint64_t key)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("BLF is not ready or not launch!");
+        return NULL;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/blf/blf.cc:395]: not formatted

File: src/supplemental/nanolib/blf/blf.cc [hunk 5]
```diff
@@ -400,6 +411,10 @@ blf_find(uint64_t key)
 const char **
 blf_find_span(uint64_t start_key, uint64_t end_key, uint32_t *size)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("BLF is not ready or not launch!");
+        return NULL;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/blf/blf.cc:417]: not formatted

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 6]
```diff
@@ -208,6 +208,10 @@ parquet_object_free(parquet_object *elem)
 int
 parquet_write_batch_async(parquet_object *elem)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("Parquet is not ready or not launch!");
+        return -1;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:214]: not formatted

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 7]
```diff
@@ -224,6 +228,10 @@ parquet_write_batch_async(parquet_object *elem)
 
 int  parquet_write_batch_tmp_async(parquet_object *elem)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("Parquet is not ready or not launch!");
+        return -1;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:234]: not formatted

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 8]
```diff
@@ -602,6 +610,10 @@ compare_callback_span(void *name, uint64_t low, uint64_t high)
 const char *
 parquet_find(uint64_t key)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("Parquet is not ready or not launch!");
+        return NULL;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:616]: not formatted

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 9]
```diff
@@ -620,6 +632,10 @@ parquet_find(uint64_t key)
 const char **
 parquet_find_span(uint64_t start_key, uint64_t end_key, uint32_t *size)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("Parquet is not ready or not launch!");
+        return NULL;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:638]: not formatted

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 10]
```diff
@@ -913,9 +929,13 @@ vector<parquet_data_packet *>
 parquet_find_data_packet(
     conf_parquet *conf, char *filename, vector<uint64_t> keys)
 {
+	vector<parquet_data_packet *> ret_vec;
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("Parquet is not ready or not launch!");
+        return ret_vec;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:936]: not formatted

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 11]
```diff
@@ -940,6 +960,10 @@ parquet_find_data_packet(
 parquet_data_packet *
 parquet_find_data_packet(conf_parquet *conf, char *filename, uint64_t key)
 {
+    if (g_conf == NULL || g_conf->enable == false) {
+        log_error("Parquet is not ready or not launch!");
+        return NULL;
+    }
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:966]: not formatted

File: src/supplemental/nanolib/ringbuffer/ringbuffer.c [hunk 12]
```diff
@@ -316,11 +317,13 @@ static parquet_object *init_parquet_object(ringBuffer_t *rb, ringBufferFile_t *f
...
 	for (unsigned int i = 0; i < rb->size; i++) {
 		keys[i] = rb->msgs[i].key;
-		smsgs[i] = rb->msgs[i].data;
 		darray[i] = nng_msg_payload_ptr((nng_msg *)rb->msgs[i].data);
 		dsize[i] = nng_msg_len((nng_msg *)rb->msgs[i].data) -
 		        (nng_msg_payload_ptr((nng_msg *)rb->msgs[i].data) -
 				(uint8_t *)nng_msg_body((nng_msg *)rb->msgs[i].data));
+
+	    nng_msg_clone((nng_msg*)rb->msgs[i].data);
+	    smsgs[i] = rb->msgs[i].data;
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/ringbuffer/ringbuffer.c:326]: not formatted

**Review by wanghaEMQ — COMMENTED** [review]
Seems a lots of places not be formatted

**Comment by wanghaEMQ** [comment]
Donot be lazy. Fix the wrong configurations in your vimrc

**Comment by xinyi-xs** [comment]
> Donot be lazy. Fix the wrong configurations in your vimrc

THX for your advice.

**Review by RanMaoyi — APPROVED** [review]

**Comment by RanMaoyi** [comment]
LGTM

**Pull request merged**